### PR TITLE
[SPARK-35820][SQL] Support Cast between different field DayTimeIntervalType

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
 Support Cast between different field DayTimeIntervalType

### Why are the changes needed?
Make user convenient to get different field DayTimeIntervalType

### Does this PR introduce _any_ user-facing change?
User can call cast DayTimeIntervalType(DAY, SECOND) to DayTimeIntervalType(DAY, MINUTE) etc

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @MaxGekk

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by MaxGekk** [comment]
@AngersZhuuuu Could resolve conflicts, please.

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu Could resolve conflicts, please.

Done

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 1]
```diff
@@ -664,6 +665,29 @@ class CastSuite extends CastSuiteBase {
     }
   }
 
+  test("SPARK-35820: Support cast DayTimeIntervalType in different fields") {
+    val ymPositive = Literal.create(
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:669]: `ymPositive`, I think `ym` shold be replaced. Also fix below values.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 2]
```diff
@@ -664,6 +665,29 @@ class CastSuite extends CastSuiteBase {
...
+      Duration.ofSeconds(12345678L, 123456789),
+      DayTimeIntervalType(DAY, SECOND))
+    val ymNegative = Literal.create(
+      Duration.ofSeconds(-12345678L, -123456789),
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:673]: You can store `Duration.ofSeconds(12345678L, 123456789)` to a val, and call `.negated()`

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: > You can store `Duration.ofSeconds(12345678L, 123456789)` to a val, and call `.negated()`

Done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: > `ymPositive`, I think `ym` shold be replaced. Also fix below values.

DOne

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 3]
```diff
@@ -664,6 +665,25 @@ class CastSuite extends CastSuiteBase {
     }
   }
 
+  test("SPARK-35820: Support cast DayTimeIntervalType in different fields") {
+    val ym = Duration.ofSeconds(12345678L, 123456789)
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:669]: This is not year-month. Please, rename `ym`.

Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:669]: > This is not year-month. Please, rename `ym`.

Done, sorry for miss this

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 4]
```diff
@@ -664,6 +665,25 @@ class CastSuite extends CastSuiteBase {
...
+      (DayTimeIntervalType(HOUR, HOUR), 12344400000000L, -12344400000000L),
+      (DayTimeIntervalType(HOUR, MINUTE), 12345660000000L, -12345660000000L),
+      (DayTimeIntervalType(HOUR, SECOND), 12345678123456L, -12345678123457L),
+      (DayTimeIntervalType(MINUTE, MINUTE), 12345660000000L, -12345660000000L),
+      (DayTimeIntervalType(MINUTE, SECOND), 12345678123456L, -12345678123457L),
+      (DayTimeIntervalType(SECOND, SECOND), 12345678123456L, -12345678123457L))
+      .foreach { case (dt, positive, negative) =>
+        checkEvaluation(cast(Literal.create(dt, DayTimeIntervalType(DAY, SECOND)), dt), positive)
+        checkEvaluation(
+          cast(Literal.create(dt.negated(), DayTimeIntervalType(DAY, SECOND)), dt), negative)
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:683]: You have 2 different dt in the same scope. Rename `val dt = Duration.ofSeconds(12345678L, 123456789)` to `duration`.

**Review by MaxGekk — APPROVED** [review]
Waiting for GA results.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master. The failed GA is not related to the changes. I ran tests (including new test) locally.
Thank you, @AngersZhuuuu .

**Pull request closed without merging**